python/pyiceberg/io/pyarrow.py

```diff
@@ -612,14 +612,40 @@ def _get_field_doc(field: pa.Field) -> Optional[str]:
 
 
 class _ConvertToIceberg(PyArrowSchemaVisitor[Union[IcebergType, Schema]]):
+    def __init__(self, expected_schema: Optional[Schema] = None):
+        self.expected_schema = expected_schema
+
+    def cast_if_needed(self, field_id: int, field_type: IcebergType) -> IcebergType:
```

Comment:
An alternative solution (a simpler one) may be adding allowing promoting from `FixedType(16)` to `UUIDType` here:https://github.com/apache/iceberg/blob/3db4e896587d95318a690979e97bab55db250e23/python/pyiceberg/schema.py#L1308-L1322

by adding 
```python
@promote.register(FixedType)
def _(file_type: FixedType, read_type: IcebergType) -> IcebergType:
    if isinstance(read_type, UUIDType) and len(file_type) == 16:
        return read_type
    else:
        raise ResolveError(f"Cannot promote an {file_type} to {read_type}")
```

My concern here is that the added promotion is not listed as a valid type promotions in [spec](https://iceberg.apache.org/spec/#schema-evolution and the effect of the added promotion may be too large. So currently I chose to narrow the fix to the visitor. Looking forward to hearing more thoughts on this.

Reply:
Hey @JonasJ-ap, first of all; thanks for catching this! 👏🏻 

Looking at it, I think adding promotions in there is the right thing to do. This way we have the correct reader three and don't have to check any types down the line.

Reply:
I'm able to reproduce this with Trino as well:
![image](https://user-images.githubusercontent.com/1134248/236166112-da4551e2-6d0a-4d46-97c8-b2a71b964619.png)

Reply:
For context, I was afraid that this would also affect other logical types, such as dates. But that seems to work, and PyArrow directly reads it as a `date32`, and no promotion is needed. Digging into it, it looks like the source of the bug is upstream in Arrow, where it doesn't pick up the UUID logical type on the Parquet field.

Reply:
Thank you so much for the explanation and test. I've changed the PR to add the promotion from FixedType[16] to UUIDType instead